src/NServiceBus.Core/Routing/SpecificInstanceDistributionPolicy.cs

```diff
@@ -0,0 +1,42 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Routing;
+
+    class SpecificInstanceDistributionPolicy : IDistributionPolicy
+    {
+        readonly string specificInstance;
+
+        public SpecificInstanceDistributionPolicy(string specificInstance)
+        {
+            this.specificInstance = specificInstance;
+        }
+
+        public DistributionStrategy GetDistributionStrategy(string endpointName)
+        {
+            return new SpecificInstanceDistributionStrategy(specificInstance);
+        }
+
+        class SpecificInstanceDistributionStrategy : DistributionStrategy
+        {
+            public SpecificInstanceDistributionStrategy(string specificInstance)
+            {
+                this.specificInstance = specificInstance;
+            }
+
+            public override IEnumerable<UnicastRoutingTarget> SelectDestination(IList<UnicastRoutingTarget> allInstances)
+            {
+                var target = allInstances.FirstOrDefault(t => t.Instance != null && t.Instance.Discriminator == specificInstance);
+                if (target == null)
+                {
+                    throw new Exception($"Specified instance {specificInstance} has not been configured in the routing tables.");
+                }
+                yield return target;
```

Comment:
@danielmarbach @bording  Although I really admire the attention to the details but I think you got a bit too far. We should optimize for readability first, and then make the core perform fast *enough*. We are not in business of microsecond-level latencies (yet). Returning `IEnumerable` and using `yield` is the most clean approach in terms of communicating the intent of the code and is also the most concise. Unless the cost is really prohibiting (e.g. like using private reflection) there is no point in optimizing for CPU cycles. It is far better to optimize for less `if` statements.